Allow non-explicit targets to be filtered out by semgrep-core

mjambon opened the pull request:
* Target files discovered in folders by semgrep are now passed to semgrep-core as "filterable targets" via a new option. Semgrep-core will then filter out whatever it thinks should be filtered out, such as for example `.min.js` files.
* Filtering targets by size is now delegated to semgrep-core

Overall these changes result in strictly more target filtering than before.

PR checklist:
- [x] Documentation is up-to-date
- [x] Changelog is up-to-date
- [x] Change has no security implications

Commit:
Add '-target' to specify explicit targets.

Commit:
Add option -filterable_target_file for the semgrep wrapper to pass a
long list of filterable targets.

Commit:
Pass explicit and filterable targets separately to semgrep-core.
Allows targets such as minified files discovered in target folders to be
skipped by semgrep-core.

Commit:
Update changelog

Commit:
Remove duplicate entry for language 'html'

Commit:
Fix type errors in tests

Commit:
Restore mypy workaround

mjambon commented on semgrep/semgrep/target_manager_extensions.py:
```diff
@@ -62,30 +73,64 @@
+    Language.TYPESCRIPT: True,
+    Language.VUE: False,
+    Language.YAML: False,
 }
```
This is for `lang_is_script`, which is unused for now. I did that earlier, but didn't finish because it's too much for one PR.

mjambon commented on semgrep/semgrep/target_manager.py:
```diff
@@ -6,27 +6,83 @@
 from pathlib import Path
 from typing import Collection
 from typing import Dict
-from typing import FrozenSet
```
Now everything uses `Set` for simplicity and to avoid unnecessary conversions. I wanted to use `FrozenSet` for everything but there was a set that was being constructed one element at a time and it would have been annoying to change it.

mjambon commented on semgrep/semgrep/target_manager.py:
```diff
@@ -6,27 +6,83 @@
+    elif lang in {"none", "generic"}:
+        return [FileExtension("*")]
+    else:
+        raise _UnknownLanguageError(f"Unsupported Language: {lang}")
```
This is weird. I need to clean it up.

mjambon replied:
At some point I was working with an old version of `target_manager.py`. I think I fixed most of the damage now.

Commit:
Undo a bunch of accidental reversions to an older version of
target_manager.py

Commit:
Fix misconceptions about --include, and always pass extensionless
executables to semgrep-core.

Commit:
Fix type errors in tests

aryx commented:
any idea about the slowdown? Noise?

aryx commented on semgrep/semgrep/target_manager_extensions.py:
```diff
@@ -13,47 +13,58 @@
+# Extensions that are permissible for a given language. Semgrep-core will
+# perform its own filtering to check that the file is indeed in the correct
+# language before attempting full parsing. Additionally, all executable
+# files without an extension will also be passed to semgrep-core.
```
What? Is this just for Bash? Should we restrict this to bash then?

IagoAbal replied:
If this wasn't done before, it may be the reason for the slow downs. Perhaps we should not do it by default and require the user to explicitly enable it via a flag?

mjambon replied:
> What? Is this just for Bash? Should we restrict this to bash then?

I'm trying to remove duplicate efforts, with a preference for ocaml over python.

This is for all the languages that can be run as scripts, which is about half of them. What would we be a benefit of doing this only when the requested language is bash?

aryx reviewed (approved):
lgtm

Commit:
Update test expectations for test_max_bytes since filtering by size is
now delegated to semgrep-core.

IagoAbal commented:
Those +10s slowdowns are probably not noise. It's a bit weird, I would expect anything but slowdowns since supposedly we will be analyzing fewer (or worst case, the same) files.

IagoAbal commented:
We definitely should check what are the before/after lists of files analyzed by semgrep-core on those now slower benchmarks. My wild guess after looking at the PR is that we may be passing files without extension to semgrep-core to analyze that previously were ignored.

Commit:
For now, comment out section that looks for executable scripts

Commit:
Resolve targets into absolute paths

Commit:
Restore functionality of the test function for comparing sets of paths
so that it resolves relative paths into absolute ones.

Commit:
In tests, partially replace target paths containing /tmp instead of
removing them completely.

IagoAbal commented:
Presumably after commit `4a4490b0d73 ("For now, comment out section that looks for executable scripts")` it seems that we no longer have that perf regression with _coinbase_. It would be nice to know why that slow down... are there so many of such files? Or are they very large? Is analyzing the files that is slow or is the heuristic to check whether we should analyze them?

Commit:
Give up on the idea of making target paths absolute due to complications
with test expectations. Left comments so it doesn't have to be
rediscovered over and over again.

mjambon commented:
I figured why the test were failing. I had incorrectly modified a function that was comparing relative paths and absolute paths without realizing it. It turns out we don't need absolute paths, and the test we have don't like absolute paths at all. In order for semgrep to represent targets as absolute paths we'd need to change our testing infrastructure, which may not be worth it. Anyway, I learned a lot and left notes and corrections about the things I discovered along the way.

Also, I commented out the code that would include executable scripts. I'll enable it in another PR when I get back to it.

mjambon commented:
@IagoAbal
> Presumably after commit 4a4490b0d73 ("For now, comment out section that looks for executable scripts") it seems that we no longer have that perf regression with coinbase. It would be nice to know why that slow down... are there so many of such files? Or are they very large? Is analyzing the files that is slow or is the heuristic to check whether we should analyze them?

I'm not expecting repos to have such a large number of executable scripts that they introduce a significant slowdown. The current approach as implemented in semgrep-core (Guess_lang module) is that given a language, we tell whether a file (executable, no extension) is in that language. In theory, it could be costly. In practice, I'm not sure. Maybe that's the problem we're seeing with the coinbase repo. If we have a problem because of n languages each inspecting the same script to determine if it's a valid target, we could start thinking of a classification mechanism: given a target, determine which language(s) it's in by inspecting it just once, instead of n times.

Anyway, this is now commented out. I'm planning to get back to this next week after I make some progress on bash.

mjambon commented:
I'm currently re-running the tests locally with `--snapshot-update` because some paths were updated. When the tests pass, this PR should be good to merge.

Commit:
Distinguish base_path from absolute_base_path.

Commit:
Update some test expectations (paths in /tmp now showing up but
truncated)

Commit:
Update expectations for explicit targets with no extension,
which are no longer treated as having an unknown extension.
